// @coreui/angular-chartjs/fesm2022/coreui-angular-chartjs.mjs
import * as i0 from '@angular/core';
import { inject, NgZone, Renderer2, ChangeDetectorRef, input, booleanAttribute, numberAttribute, linkedSignal, output, viewChild, computed, afterRenderEffect, effect, untracked, ChangeDetectionStrategy, Component, NgModule } from '@angular/core';
import { merge } from 'lodash-es';
import { Chart, registerables } from 'chart.js';
import { customTooltips } from '@coreui/chartjs';

Chart.register(...registerables);
let nextId = 0;
class ChartjsComponent {
    //
    static ngAcceptInputType_redraw;
    ngZone = inject(NgZone);
    renderer = inject(Renderer2);
    changeDetectorRef = inject(ChangeDetectorRef);
    /**
     * Enables custom html based tooltips instead of standard tooltips.
     * @return boolean
     * @default true
     */
    customTooltips = input(true, { ...(ngDevMode ? { debugName: "customTooltips" } : /* istanbul ignore next */ {}), transform: booleanAttribute });
    /**
     * The data object that is passed into the Chart.js chart (more info).
     * @returns ChartData
     * @default { labels: [], datasets: [] }
     */
    data = input(...(ngDevMode ? [undefined, { debugName: "data" }] : /* istanbul ignore next */ []));
    /**
     * A fallback when the canvas cannot be rendered. Can be used for accessible chart descriptions.
     */
    // @Input() fallbackContent?: TemplateRef<any>;
    /**
     * Height attribute applied to the rendered canvas.
     * @return number | undefined
     * @default null
     */
    height = input(null, { ...(ngDevMode ? { debugName: "height" } : /* istanbul ignore next */ {}), transform: (value) => numberAttribute(value, undefined) });
    /**
     * ID attribute applied to the rendered canvas.
     * @returns string
     * @default 'c-chartjs-' + nextId
     */
    idInput = input(`c-chartjs-${nextId++}`, { ...(ngDevMode ? { debugName: "idInput" } : /* istanbul ignore next */ {}), alias: 'id' });
    get id() {
        return this.idInput();
    }
    /**
     * The options object that is passed into the Chart.js chart.
     * @returns ChartOptions | undefined
     * @default {}
     */
    optionsInput = input({}, { ...(ngDevMode ? { debugName: "optionsInput" } : /* istanbul ignore next */ {}), alias: 'options' });
    options = linkedSignal(this.optionsInput, ...(ngDevMode ? [{ debugName: "options" }] : /* istanbul ignore next */ []));
    /**
     * The plugins array that is passed into the Chart.js chart
     * @returns Plugin[]
     * @default []
     */
    plugins = input([], ...(ngDevMode ? [{ debugName: "plugins" }] : /* istanbul ignore next */ []));
    /**
     * If true, will tear down and redraw chart on all updates.
     * @returns boolean
     * @default false
     */
    redraw = input(false, { ...(ngDevMode ? { debugName: "redraw" } : /* istanbul ignore next */ {}), transform: booleanAttribute });
    /**
     * Chart.js chart type.
     * @returns {'line' | 'bar' | 'radar' | 'doughnut' | 'polarArea' | 'bubble' | 'pie' | 'scatter'}
     * @default 'bar'
     */
    type = input('bar', ...(ngDevMode ? [{ debugName: "type" }] : /* istanbul ignore next */ []));
    /**
     * Width attribute applied to the rendered canvas.
     * @return number | undefined
     * @default null
     */
    width = input(null, { ...(ngDevMode ? { debugName: "width" } : /* istanbul ignore next */ {}), transform: (value) => numberAttribute(value, undefined) });
    /**
     * Put the chart into the wrapper div element.
     * @return boolean
     * @default true
     */
    wrapper = input(true, { ...(ngDevMode ? { debugName: "wrapper" } : /* istanbul ignore next */ {}), transform: booleanAttribute });
    /** Event emitted on click of a chart element. Contains an array of the clicked elements.
     * @returns InteractionItem[]
     * @see https://www.chartjs.org/docs/latest/developers/api.html#getelementsateventformode
     */
    getDatasetAtEvent = output();
    /** Event emitted on click of a chart element. Contains an array of the clicked element.
     * @returns InteractionItem[]
     * @see https://www.chartjs.org/docs/latest/developers/api.html#getelementsateventformode
     */
    getElementAtEvent = output();
    /** Event emitted on click of a chart element. Contains an array of the clicked elements.
     * @returns InteractionItem[]
     * @see https://www.chartjs.org/docs/latest/developers/api.html#getelementsateventformode
     */
    getElementsAtEvent = output();
    /** Emits the Chart.js chart instance after it is created. Can be used to access Chart.js directly on the chart instance.
     * @returns ChartJS | undefined
     * @see https://www.chartjs.org/docs/latest/developers/api.html#chart-instance-methods
     */
    chartRef = output();
    canvasElement = viewChild.required('canvasElement');
    chart;
    ctx = computed(() => this.canvasElement()?.nativeElement.getContext('2d'), ...(ngDevMode ? [{ debugName: "ctx" }] : /* istanbul ignore next */ []));
    constructor() {
        afterRenderEffect({
            read: () => {
                const ctx = this.ctx();
                if (ctx && !this.chart) {
                    this.chartRender();
                }
            }
        });
        effect(() => {
            const data = this.data();
            untracked(() => {
                if (this.chart && data) {
                    this.chartUpdate();
                }
            });
        });
    }
    ngOnDestroy() {
        this.chartDestroy();
    }
    handleClick($event) {
        if (!this.chart) {
            return;
        }
        this.emitDatasetAtEvent($event);
        this.emitElementAtEvent($event);
        this.emitElementsAtEvent($event);
    }
    emitDatasetAtEvent($event) {
        const datasetAtEvent = this.chart?.getElementsAtEventForMode($event, 'dataset', { intersect: true }, false);
        if (datasetAtEvent) {
            this.getDatasetAtEvent.emit(datasetAtEvent);
        }
    }
    emitElementAtEvent($event) {
        const elementAtEvent = this.chart?.getElementsAtEventForMode($event, 'nearest', { intersect: true }, false);
        if (elementAtEvent) {
            this.getElementAtEvent.emit(elementAtEvent);
        }
    }
    emitElementsAtEvent($event) {
        const elementsAtEvent = this.chart?.getElementsAtEventForMode($event, 'index', { intersect: true }, false);
        if (elementsAtEvent) {
            this.getElementsAtEvent.emit(elementsAtEvent);
        }
    }
    chartDestroy() {
        this.chart?.destroy();
        this.chartRef.emit(undefined);
    }
    chartRender() {
        const canvasElement = this.canvasElement();
        if (!canvasElement?.nativeElement || !this.ctx() || this.chart) {
            return;
        }
        this.ngZone.runOutsideAngular(() => {
            const config = this.chartConfig();
            const ctx = this.ctx();
            if (config && ctx) {
                this.chart = new Chart(ctx, config);
                this.ngZone.run(() => {
                    this.renderer.setStyle(canvasElement.nativeElement, 'display', 'block');
                    this.changeDetectorRef.markForCheck();
                    this.chartRef.emit(this.chart);
                });
            }
        });
    }
    chartUpdate() {
        if (!this.chart) {
            return;
        }
        if (this.redraw()) {
            this.chartDestroy();
            this.chartRender();
            return;
        }
        const config = this.chartConfig();
        if (this.options()) {
            Object.assign(this.chart.options ?? {}, config.options ?? {});
        }
        if (!this.chart.config.data) {
            this.chart.config.data = { ...config.data };
            this.chartUpdateOutsideAngular();
        }
        if (this.chart) {
            Object.assign(this.chart.config.options ?? {}, config.options ?? {});
            Object.assign(this.chart.config.plugins ?? [], config.plugins ?? []);
            Object.assign(this.chart.config.data, config.data);
        }
        this.chartUpdateOutsideAngular();
    }
    chartUpdateOutsideAngular() {
        setTimeout(() => {
            this.ngZone.runOutsideAngular(() => {
                try {
                    if (this.chart?.canvas) {
                        this.chart?.update();
                    }
                }
                catch (error) {
                    console.warn('Error on chart.update():', error);
                }
                this.ngZone.run(() => {
                    this.changeDetectorRef.markForCheck();
                });
            });
        });
    }
    chartToBase64Image() {
        return this.chart?.toBase64Image();
    }
    chartDataConfig = computed(() => {
        const { labels, datasets } = { ...this.data() };
        return {
            labels: labels ?? [],
            datasets: datasets ?? []
        };
    }, ...(ngDevMode ? [{ debugName: "chartDataConfig" }] : /* istanbul ignore next */ []));
    chartOptions = computed(() => this.options() ?? {}, ...(ngDevMode ? [{ debugName: "chartOptions" }] : /* istanbul ignore next */ []));
    chartConfig = computed(() => {
        this.chartCustomTooltips();
        return {
            data: this.chartDataConfig(),
            options: this.chartOptions(),
            plugins: this.plugins(),
            type: this.type()
        };
    }, ...(ngDevMode ? [{ debugName: "chartConfig" }] : /* istanbul ignore next */ []));
    chartCustomTooltips() {
        if (!this.customTooltips()) {
            return;
        }
        const options = this.options();
        const plugins = options?.plugins;
        const tooltip = options?.plugins?.tooltip;
        untracked(() => {
            this.options.set(merge({
                ...options,
                plugins: {
                    ...plugins,
                    tooltip: {
                        ...tooltip,
                        enabled: false,
                        mode: 'index',
                        position: 'nearest',
                        external: customTooltips
                    }
                }
            }));
        });
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.10", ngImport: i0, type: ChartjsComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.2.0", version: "21.2.10", type: ChartjsComponent, isStandalone: true, selector: "c-chart", inputs: { customTooltips: { classPropertyName: "customTooltips", publicName: "customTooltips", isSignal: true, isRequired: false, transformFunction: null }, data: { classPropertyName: "data", publicName: "data", isSignal: true, isRequired: false, transformFunction: null }, height: { classPropertyName: "height", publicName: "height", isSignal: true, isRequired: false, transformFunction: null }, idInput: { classPropertyName: "idInput", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, optionsInput: { classPropertyName: "optionsInput", publicName: "options", isSignal: true, isRequired: false, transformFunction: null }, plugins: { classPropertyName: "plugins", publicName: "plugins", isSignal: true, isRequired: false, transformFunction: null }, redraw: { classPropertyName: "redraw", publicName: "redraw", isSignal: true, isRequired: false, transformFunction: null }, type: { classPropertyName: "type", publicName: "type", isSignal: true, isRequired: false, transformFunction: null }, width: { classPropertyName: "width", publicName: "width", isSignal: true, isRequired: false, transformFunction: null }, wrapper: { classPropertyName: "wrapper", publicName: "wrapper", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { getDatasetAtEvent: "getDatasetAtEvent", getElementAtEvent: "getElementAtEvent", getElementsAtEvent: "getElementsAtEvent", chartRef: "chartRef" }, host: { properties: { "class.chart-wrapper": "wrapper()", "style.height.px": "height()", "style.width.px": "width()" } }, viewQueries: [{ propertyName: "canvasElement", first: true, predicate: ["canvasElement"], descendants: true, isSignal: true }], exportAs: ["cChart"], ngImport: i0, template: "<canvas\n  #canvasElement\n  (click)=\"handleClick($event)\"\n  [height]=\"height() || null\"\n  [id]=\"id\"\n  [width]=\"width() || null\"\n  role=\"img\"\n  style=\"display: none;\"\n>\n  <ng-content />\n  <!--  <ng-container *ngTemplateOutlet=\"fallbackContent\"/>-->\n</canvas>\n", styles: [":host.chart-wrapper{display:block}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.10", ngImport: i0, type: ChartjsComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c-chart', exportAs: 'cChart', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[class.chart-wrapper]': 'wrapper()',
                        '[style.height.px]': 'height()',
                        '[style.width.px]': 'width()'
                    }, template: "<canvas\n  #canvasElement\n  (click)=\"handleClick($event)\"\n  [height]=\"height() || null\"\n  [id]=\"id\"\n  [width]=\"width() || null\"\n  role=\"img\"\n  style=\"display: none;\"\n>\n  <ng-content />\n  <!--  <ng-container *ngTemplateOutlet=\"fallbackContent\"/>-->\n</canvas>\n", styles: [":host.chart-wrapper{display:block}\n"] }]
        }], ctorParameters: () => [], propDecorators: { customTooltips: [{ type: i0.Input, args: [{ isSignal: true, alias: "customTooltips", required: false }] }], data: [{ type: i0.Input, args: [{ isSignal: true, alias: "data", required: false }] }], height: [{ type: i0.Input, args: [{ isSignal: true, alias: "height", required: false }] }], idInput: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], optionsInput: [{ type: i0.Input, args: [{ isSignal: true, alias: "options", required: false }] }], plugins: [{ type: i0.Input, args: [{ isSignal: true, alias: "plugins", required: false }] }], redraw: [{ type: i0.Input, args: [{ isSignal: true, alias: "redraw", required: false }] }], type: [{ type: i0.Input, args: [{ isSignal: true, alias: "type", required: false }] }], width: [{ type: i0.Input, args: [{ isSignal: true, alias: "width", required: false }] }], wrapper: [{ type: i0.Input, args: [{ isSignal: true, alias: "wrapper", required: false }] }], getDatasetAtEvent: [{ type: i0.Output, args: ["getDatasetAtEvent"] }], getElementAtEvent: [{ type: i0.Output, args: ["getElementAtEvent"] }], getElementsAtEvent: [{ type: i0.Output, args: ["getElementsAtEvent"] }], chartRef: [{ type: i0.Output, args: ["chartRef"] }], canvasElement: [{ type: i0.ViewChild, args: ['canvasElement', { isSignal: true }] }] } });

class ChartjsModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.10", ngImport: i0, type: ChartjsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "21.2.10", ngImport: i0, type: ChartjsModule, imports: [ChartjsComponent], exports: [ChartjsComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "21.2.10", ngImport: i0, type: ChartjsModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.10", ngImport: i0, type: ChartjsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        ChartjsComponent
                    ],
                    exports: [
                        ChartjsComponent
                    ]
                }]
        }] });

/*
 * Public API Surface of coreui-angular-chartjs
 */

/**
 * Generated bundle index. Do not edit.
 */

export { ChartjsComponent, ChartjsModule };
